### PR TITLE
fix on Julia nightly

### PR DESCRIPTION
**Pull request by simeonschaub**
see https://github.com/JuliaLang/julia/pull/38835 and https://github.com/JuliaLang/julia/pull/38136

### CODE REVIEW

**Review by pfitzseb — DISMISSED** [review]
LGTM, but @ZacLN should probably take a look at this too.

**Comment by simeonschaub** [comment]
Hmm, ok, I am now not quite sure how `FakeTypeName` is supposed to work. Is it expected that something like `Array{Int}` just drops the parameters for the `DataType` `Array{Int,TypeName(:N)}` completely? It seems to be intentional because `FakeTypeName` is called with `justname=true`.

**Review by ZacLN — APPROVED** [review]

**Comment by simeonschaub** [comment]
Bump. What should be done about the tests here?

File: src/faketypes.jl [hunk 1]
```diff
@@ -120,3 +123,47 @@ Base.:(==)(a::FakeTypeVar, b::FakeTypeVar) = a.lb == b.lb && a.name == b.name &&
...
+        N
+        FakeTypeofVararg() = new()
+        FakeTypeofVararg(T) = new(T)
+        FakeTypeofVararg(T, N) = new(T, N)
+    end
+    function FakeTypeofVararg(va::Core.TypeofVararg)
+        if isdefined(va, :N)
+            FakeTypeofVararg(FakeTypeName(va.T), va.N)
+        elseif isdefined(va, :T)
+            FakeTypeofVararg(FakeTypeName(va.T))
```
Comment (simeonschaub) [inline src/faketypes.jl:139]: Is this more correct?
```suggestion
            FakeTypeofVararg(FakeTypeName(va.T; justname=true), va.N)
        elseif isdefined(va, :T)
            FakeTypeofVararg(FakeTypeName(va.T; justname=true))
```

**Comment by musm** [comment]
bump, would love to be able to use this on 1.7

**Comment by ZacLN** [comment]
> Hmm, ok, I am now not quite sure how `FakeTypeName` is supposed to work. Is it expected that something like `Array{Int}` just drops the parameters for the `DataType` `Array{Int,TypeName(:N)}` completely? It seems to be intentional because `FakeTypeName` is called with `justname=true`.

Yes, in it's current state this isn't intended to fully describe the types. Parameters are dropped except where at the first level of a DataType. For example given `abstract type T{S} end` we would have a full representaiton of `T{Bool}` but not of `T{T{Bool}}` and so `FakeTypeName(T{T{SomeType}}) == FakeTypeName(T{T{SomeOtherType}})` is expected. This is to prevent storing 'massive' datatypes - there are some huge ones in LinearAlgebra. Given this, I don't think the tests are quite appropriate (i.e. they're asking too much) and happy for them to be dropped.

I've not had a look at the new approach to Varargs and a different heuristic may well be appropriate.

**Comment by simeonschaub** [comment]
I still kept most of the tests, but excluded some in earlier Julia versions. I hope they are not too strict now, but my thought was that they can always be loosened if they become a problem. This also now passes on `justname=true` for all type parameters of `Vararg` to be consistent with the previous behavior.

> This is to prevent storing 'massive' datatypes - there are some huge ones in LinearAlgebra.

Do you mean things like `StridedArray`? Maybe we could piggyback off the type alias printing added to Base in https://github.com/JuliaLang/julia/pull/36107 here, so we only store the alias if there is a matching one.

**Review by ZacLN — APPROVED** [review]

**Comment by ZacLN** [comment]
Thanks, looks good. Yes exactly that, I'll have a look at the PR (thanks for the ref).

**Comment by simeonschaub** [comment]
Good to merge?

**Comment by musm** [comment]
Would also be great to have a tag with the fix.

**Comment by ZacLN** [comment]
Yes, staticlint will need small adjustments.

File: .github/workflows/jlpkgbutler-ci-master-workflow.yml [hunk 2]
```diff
@@ -10,7 +10,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        julia-version: ['1.0', '1.1', '1.2', '1.3', '1.4', '1.5']
+        julia-version: ['1.0', '1.1', '1.2', '1.3', '1.4', '1.5', 'nightly']
```
Comment (davidanthoff) [inline .github/workflows/jlpkgbutler-ci-master-workflow.yml:13]: We should generally not edit these files manually, they are managed by package butler and any change we make manually here will be undone by package butler down the road.

Comment (simeonschaub) [reply .github/workflows/jlpkgbutler-ci-master-workflow.yml:13]: Oops, sorry! Didn't realize they were automatically generated.

Comment (davidanthoff) [reply .github/workflows/jlpkgbutler-ci-master-workflow.yml:13]: No harm done, it will just disappear at some point again :)

**Pull request merged**